ddl/ddl_api.go

```diff
@@ -3519,32 +3526,111 @@ func (d *ddl) CreateIndex(ctx sessionctx.Context, ti ast.Ident, keyType ast.Inde
 	}
 
 	tblInfo := t.Meta()
+	hiddenCols := make([]*model.ColumnInfo, 0, len(indexPartSpecifications))
+	// build hidden columns if necessary
+	for i, idxPart := range indexPartSpecifications {
+		if idxPart.Expr != nil {
+			idxPart.Column = &ast.ColumnName{Name: model.NewCIStr(fmt.Sprintf("_v$_%s_%d", indexName, i))}
+			idxPart.Length = types.UnspecifiedLength
+			// The index part is an expression, prepare a hidden column for it.
+			if len(idxPart.Column.Name.L) > mysql.MaxColumnNameLength {
+				// TODO: refine the error message
+				return ErrTooLongIdent.GenWithStackByArgs("hidden column")
+			}
+			// // TODO: refine the error message
+			if err := checkIllegalFn4GeneratedColumn("expression index", idxPart.Expr); err != nil {
+				return errors.Trace(err)
+			}
+
+			var sb strings.Builder
+			restoreFlags := format.RestoreStringSingleQuotes | format.RestoreKeyWordLowercase | format.RestoreNameBackQuotes |
+				format.RestoreSpacesAroundBinaryOperation
+			restoreCtx := format.NewRestoreCtx(restoreFlags, &sb)
+			sb.Reset()
+			err = idxPart.Expr.Restore(restoreCtx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			expr, err := expression.RewriteSimpleExprWithTableInfo(ctx, tblInfo, idxPart.Expr)
+			if err != nil {
+				// TODO: refine the error message
+				return err
+			}
+			if _, ok := expr.(*expression.Column); ok {
+				return ErrFunctionalIndexOnField
+			}
+
+			colInfo := &model.ColumnInfo{
+				Name:                idxPart.Column.Name,
+				GeneratedExprString: sb.String(),
+				GeneratedStored:     false,
+				Version:             model.CurrLatestColumnInfoVersion,
+				Dependences:         make(map[string]struct{}),
+				Hidden:              true,
+				FieldType:           *expr.GetType(),
+			}
+			for _, colName := range findColumnNamesInExpr(idxPart.Expr) {
+				colInfo.Dependences[colName.Name.O] = struct{}{}
+			}
+			if err = checkDependedColExist(colInfo.Dependences, t.Cols()); err != nil {
+				return errors.Trace(err)
+			}
+			idxPart.Expr = nil
+			hiddenCols = append(hiddenCols, colInfo)
+		}
+	}
```

Comment:
Extract these code to a function, `create Table ... index ...` statement need this logic too.

Reply:
Done